app/worker/worker.go

```diff
@@ -86,56 +87,123 @@ func InitAndRun(
 		_ = unsubscribe()
 	}()
 
-	var msg interface{}
-	sem := make(chan bool, concurrency)
 	quit := make(chan os.Signal, 1)
 	signal.Notify(quit, unix.SIGINT, unix.SIGTERM)
+
+	// Spawn worker pool
+	var wg = new(sync.WaitGroup)
+	wg.Add(concurrency)
+	for i := 0; i < concurrency; i++ {
+		workerID := i
+		go func() {
+			defer func() {
+				wg.Done()
+				wg.Wait()
+				// Make sure context is cancelled when all workers are down
+				cancel()
+			}()
+			wl := l.F(log.Ctx{"worker_id": workerID})
+			wCtx := log.WithContext(ctx, wl)
+
+			wl.Info("worker starting up")
+			workerMain(wCtx, channel, natsClient, dataStore)
+			wl.Info("worker shut down")
+		}()
+	}
+
+	// Run until a SIGTERM or SIGINT is received or the entire pool is down
+	select {
+	case <-quit:
+	case <-ctx.Done():
+	}
+	// Notify workers that we're done
+	close(channel)
+	// Wait for all workers to finish their current task
+	wg.Wait()
+	return nil
+}
+
+func workerMain(ctx context.Context, msgIn <-chan *natsio.Msg, nc nats.Client, ds store.DataStore) {
+	l := log.FromContext(ctx)
+	processMsgChan := make(chan *natsio.Msg, 1)
+	defer close(processMsgChan)
+	done := ctx.Done()
+
+	// workerSidecar is responsible for notifying the broker about slow workflows
+	go workerSidecar(ctx, processMsgChan)
```

Comment:
Honestly I don't get how the sidercar receives the message via processMsgChan.
All I see is a `<- nil`.

Reply:
Nice catch. I think I may have accidentally removed that line while refactoring my initial code.

Reply:
Fixed.